### ISSUE TITLE
[BUG] Cannot provide a custom ArtifactRepository plugin without implementing a custom FileStore/ModelRegistryStore

### ISSUE THREAD

**Issue by ChrisSchnablQC** [mlflow/mlflow#11706]
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Local machine

### Willingness to contribute

Yes. I can contribute a fix for this bug independently.

### MLflow version

- Client:. 2.11.3

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: macOS Sonoma 14.3/Darwin Darwin Kernel Version 23.3.0
- **Python version**:  3.12 
- **MLflow version**: 2.11.3
- **MLflow module location**: /Users/chris/mlflow-plugin-test/.pixi/envs/default/lib/python3.12/site-packages/mlflow/__init__.py
- **Tracking URI**: file-plugin://127.0.0.1:8080
- **Registry URI**: file:///...../..../mlflow-plugin-test/mlruns

## Context

I am developing a custom `ArtifactRepository` plugin by subclassing the `ArtifactRepository` class. This includes providing implementations for `log_artifact` and other required methods, and registering it as an entry point under the `file-plugin` scheme. My goal is to continue using local file storage for both `tracking.FileStore` and `model_registry.FileStore` as implementations of `tracking.AbstractStore`. According to the documentation [0], [1], [2], this configuration should be supported.

However, I have encountered several significant issues with the current implementation of MLFlow that complicate this integration, listed in order of occurrence:
1. **Model Registry and Tracking URI Conflict**: The `MlflowClient` attempts to register a `file-plugin` compatible `ModelRegistryStore` for the `tracking_client`. This is not resolved by setting `model_registry_uri`.
2. **Scheme Conflict**: Providing a `tracking.AbstractStore` implementation with the `file-plugin` scheme solves the first issue. However, using a `tracking.AbstractStore` that is different from the `ArtifactRepository` leads to conflicts. For example, during `log_artifact`, MLFlow matches the tracking repository using the same scheme, regardless of whether a custom `ModelRegistry` is implemented.
3. **Custom Model Registry Registration**: While implementing a `CustomModelRegistry` resolves some issues, it fails to register properly within the system.

## First Problem
Initially, I subclassed the `ArtifactRepository`, set the tracking URI, initialized an experiment, and logged some artifacts. However, MLFlow implicitly sets the model registry URI to the tracking URI, which is logical but problematic. Even when specifying a custom model registry URI with a supported scheme, it fails to locate the appropriate store. This behavior can be reproduced with the script below, even without a plugin installed.

## Second Problem
A temporary workaround involves redirecting `file-plugin` requests to a `FileStore` plugin using an existing implementation:
```python
class FileStore(AbstractStore):
    def __init__(self, store_uri=None, artifact_uri=None):
        self.is_plugin = True
        super().__init__()
```
However, this approach has limitations. MLFlow uses the `run_id` to retrieve the `artifact_uri` from run information, coupling the two plugins. This behavior can be demonstrated by modifying the run info `artifact_uri` to point to the remote file plugin location:
```python
class FileStore(AbstractStore):
    ...
    def get_run(self, run_id):
        run = super().get_run(run_id)
        run.info._artifact_uri = "file-plugin://"
        return run
```

## Third Problem
When attempting to register a local registry with the configuration below, the model registries fail to be added, despite successful addition of local artifact repositories:
```
[project.entry-points."mlflow.mlflow.model_registry_store"]
"file-plugin" = "mlflow_test_plugin.local_model_registry:LocalModelRegistry"
```
```python
class LocalModelRegistry(AbstractStore):
    def __init__(self, store_uri=None, tracking_uri=None):
        self.is_plugin = True
        super().__init__()
```
Debugging this via `entrypoints.py` reveals issues in the registration process.

## Conclusion
Writing a custom implementation for just the `ArtifactRepository` without having to reimplement the other backend store options, or wrapping them, or adding other hacky tricks to correctly set registry and tracking URIs should be supported out of the box. This would potentially block many developers from facing difficulties in trying to integrate different open-source plug-and-play solutions for various MLFlow components. If there are any behaviors that I did not consider or any MLFlow-idiomatic solutions, I would appreciate any pointers. I would also be willing to update the documentation to make it more transparent about what can and needs to be done to implement certain parts of a storage backend, even if that just means clarifying that certain implementations are not advised/possible within current MLFlow limitations. 

### Tracking information

```shell
System information: Darwin Darwin Kernel Version 23.3.0
Python version: 3.12.2
MLflow version: 2.11.3
MLflow module location: /.../.../mlflow-plugin-test/.pixi/envs/default/lib/python3.12/site-packages/mlflow/__init__.py
Tracking URI: file-plugin://127.0.0.1:8080
Registry URI: file:///.../.../mlflow-plugin-test/mlruns
MLflow dependencies: 
  Flask: 3.0.2
  Jinja2: 3.1.3
  alembic: 1.13.1
  click: 8.1.7
  cloudpickle: 3.0.0
  docker: 7.0.0
  entrypoints: 0.4
  gitpython: 3.1.43
  graphene: 3.3
  gunicorn: 21.2.0
  importlib-metadata: 7.1.0
  markdown: 3.6
  matplotlib: 3.8.3
  numpy: 1.26.4
  packaging: 23.2
  pandas: 2.2.1
  prometheus-flask-exporter: 0.23.0
  protobuf: 4.25.3
  pyarrow: 15.0.2
  pytz: 2024.1
  pyyaml: 6.0.1
  querystring-parser: 1.2.4
  requests: 2.31.0
  scikit-learn: 1.4.1.post1
  scipy: 1.13.0
  sqlalchemy: 2.0.29
  sqlparse: 0.4.4
```

### Code to reproduce issue

```python
import mlflow

mlflow.set_registry_uri(uri="file:///Users/chris/mlflow-plugin-test/mlruns")
mlflow.set_tracking_uri(uri="file-plugin://127.0.0.1:8080")

mlflow.doctor()
mlflow.set_experiment("test")
```

### Stack trace

```python
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
File ~/mlflow-plugin-test/.pixi/envs/default/lib/python3.12/site-packages/mlflow/tracking/registry.py:83, in StoreRegistry.get_store_builder(self, store_uri)
     82 try:
---> 83     store_builder = self._registry[scheme]
     84 except KeyError:

KeyError: 'file-plugin'

During handling of the above exception, another exception occurred:

UnsupportedModelRegistryStoreURIException Traceback (most recent call last)
Cell In[12], line 5
      2 mlflow.set_tracking_uri(uri="file-plugin://127.0.0.1:8080")
      4 mlflow.doctor()
----> 5 mlflow.set_experiment("test")

File ~/mlflow-plugin-test/.pixi/envs/default/lib/python3.12/site-packages/mlflow/tracking/fluent.py:140, in set_experiment(experiment_name, experiment_id)
    132 if (experiment_name is not None and experiment_id is not None) or (
    133     experiment_name is None and experiment_id is None
    134 ):
    135     raise MlflowException(
    136         message="Must specify exactly one of: `experiment_id` or `experiment_name`.",
    137         error_code=INVALID_PARAMETER_VALUE,
    138     )
--> 140 client = MlflowClient()
    141 if experiment_id is None:
    142     experiment = client.get_experiment_by_name(experiment_name)

File ~/mlflow-plugin-test/.pixi/envs/default/lib/python3.12/site-packages/mlflow/tracking/client.py:94, in MlflowClient.__init__(self, tracking_uri, registry_uri)
     92 final_tracking_uri = utils._resolve_tracking_uri(tracking_uri)
     93 self._registry_uri = registry_utils._resolve_registry_uri(registry_uri, tracking_uri)
---> 94 self._tracking_client = TrackingServiceClient(final_tracking_uri)

File ~/mlflow-plugin-test/.pixi/envs/default/lib/python3.12/site-packages/mlflow/tracking/_tracking_service/client.py:53, in TrackingServiceClient.__init__(self, tracking_uri)
     48 self.tracking_uri = tracking_uri
     49 # NB: Fetch the tracking store (`self.store`) upon client initialization to ensure that
     50 # the tracking URI is valid and the store can be properly resolved. We define `store` as a
     51 # property method to ensure that the client is serializable, even if the store is not
     52 # self.store
---> 53 self.store

File ~/mlflow-plugin-test/.pixi/envs/default/lib/python3.12/site-packages/mlflow/tracking/_tracking_service/client.py:57, in TrackingServiceClient.store(self)
     55 @property
     56 def store(self):
---> 57     return utils._get_store(self.tracking_uri)

File ~/mlflow-plugin-test/.pixi/envs/default/lib/python3.12/site-packages/mlflow/tracking/_tracking_service/utils.py:196, in _get_store(store_uri, artifact_uri)
    195 def _get_store(store_uri=None, artifact_uri=None):
--> 196     return _tracking_store_registry.get_store(store_uri, artifact_uri)

File ~/mlflow-plugin-test/.pixi/envs/default/lib/python3.12/site-packages/mlflow/tracking/_tracking_service/registry.py:42, in TrackingStoreRegistry.get_store(self, store_uri, artifact_uri)
     39 from mlflow.tracking._tracking_service import utils
     41 resolved_store_uri = utils._resolve_tracking_uri(store_uri)
---> 42 return self._get_store_with_resolved_uri(resolved_store_uri, artifact_uri)

File ~/mlflow-plugin-test/.pixi/envs/default/lib/python3.12/site-packages/mlflow/tracking/_tracking_service/registry.py:51, in TrackingStoreRegistry._get_store_with_resolved_uri(self, resolved_store_uri, artifact_uri)
     44 @lru_cache(maxsize=100)
     45 def _get_store_with_resolved_uri(self, resolved_store_uri, artifact_uri=None):
     46     """
     47     Retrieve the store associated with a resolved (non-None) store URI and an artifact URI.
     48     Caching is done on resolved URIs because the meaning of an unresolved (None) URI may change
     49     depending on external configuration, such as environment variables
     50     """
---> 51     builder = self.get_store_builder(resolved_store_uri)
     52     return builder(store_uri=resolved_store_uri, artifact_uri=artifact_uri)

File ~/mlflow-plugin-test/.pixi/envs/default/lib/python3.12/site-packages/mlflow/tracking/registry.py:85, in StoreRegistry.get_store_builder(self, store_uri)
     83     store_builder = self._registry[scheme]
     84 except KeyError:
---> 85     raise UnsupportedModelRegistryStoreURIException(
     86         unsupported_uri=store_uri, supported_uri_schemes=list(self._registry.keys())
     87     )
     88 return store_builder

UnsupportedModelRegistryStoreURIException:  Model registry functionality is unavailable; got unsupported URI 'file-plugin://127.0.0.1:8080' for model registry data storage. Supported URI schemes are: ['', 'file', 'databricks', 'databricks-uc', 'http', 'https', 'postgresql', 'mysql', 'sqlite', 'mssql']. See https://www.mlflow.org/docs/latest/tracking.html#storage for how to run an MLflow server against one of the supported backend storage locations.
```

### What component(s) does this bug affect?

- [X] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [X] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by serena-ruan** [comment]
@ChrisSchnablQC Did you follow https://mlflow.org/docs/latest/plugins.html#writing-your-own-mlflow-plugins for writing the plugin? entry_points are collected in setup.py file. BTW I don't get where you mentioned that specifying a model_registry_uri doesn't work if you specify the tracking_uri, could you provide code snippet and stacktrace?

**Comment by ChrisSchnablQC** [comment]
@serena-ruan Yes, I've verified that the plugin is correctly installed (in my case through a pixi/pyproject.toml setup but using a setup.py gives the same result so we can assume that for none) and registered through the entrypoints.py. 
Re BTW: Please have a look at Code to reproduce and the stack trace, this is the simplest example I found when having a using ArtifactRepository plugin with the file-plugin scheme. Let me know if you can reproduce the output.

**Comment by ChrisSchnablQC** [comment]
@serena-ruan When I debug MLFlow and set a breakpoint in entrypoints.py the plugin for the artifact repository group is registered. I've also manually checked that it exists in `from mlflow.store.artifact import artifact_repository_registry
r = artifact_repository_registry` just now. The problem also occurs when I don't use a custom plugin, but e.g. want to configure the tracking-uri as s3.

**Comment by serena-ruan** [comment]
If you set a breakpoint here https://github.com/mlflow/mlflow/blob/4374f624f92f8160471dbf8d388d117b186dee00/mlflow/tracking/_tracking_service/utils.py#L196 Does _tracking_store_registry contains 'file-plugin'?

**Comment by ChrisSchnablQC** [comment]
No, it does not, but should it? From my understanding you would register plugins that implement `mlflow.store.tracking.AbstractStore` there.

**Comment by serena-ruan** [comment]
Yes it should include the plugin, that's where the exception throws. Isn't tracking store what you want to replace?

**Comment by ChrisSchnablQC** [comment]
Yes, this is what I have been doing - see Context: 

> I am developing a custom ArtifactRepository plugin by subclassing the ArtifactRepository class. This includes providing implementations for log_artifact and other required methods, and registering it as an entry point under the file-plugin scheme. 

And this is also why is was surprised to see that MLFlow thinks I also need to add a file-store/model-registry store.

**Comment by serena-ruan** [comment]
If you still want to use the local file store for tracking and only remote storage for logging artifacts, then you'll need to do something like this:
```
import mlflow
mlflow.get_tracking_uri() # by default it's local file store # 'file:///Users/serena.ruan/Documents/test/mlruns'
mlflow.create_experiment("test", artifact_location="s3://serena-ruan-test") # I'm using s3 as an example, you should replace with your artifact repository scheme
mlflow.set_experiment("test')
with mlflow.start_run():
    mlflow.log_artifact("some_file.txt")
```
